File: backend/app/routers/portfolio.py
from fastapi import APIRouter, HTTPException

from app.models import Order, OrderRequest, Position, AccountBalance, RiskCheckResult
from app.services import futu_mock as futu

router = APIRouter(prefix="/api/portfolio", tags=["portfolio"])


@router.get("/positions", response_model=list[Position])
async def get_positions():
    return futu.get_positions()


@router.get("/balance", response_model=AccountBalance)
async def get_balance():
    return futu.get_balance()


@router.post("/risk-check", response_model=RiskCheckResult)
async def risk_check(req: OrderRequest):
    balance = futu.get_balance()
    positions = futu.get_positions()
    return futu.check_risk(req, balance, positions)


@router.post("/orders", response_model=Order, status_code=201)
async def place_order(req: OrderRequest):
    if not req.paper_trade:
        raise HTTPException(status_code=400, detail="实盘交易未启用，请设置 paper_trade=true")
    balance = futu.get_balance()
    positions = futu.get_positions()
    risk = futu.check_risk(req, balance, positions)
    if not risk.passed:
        raise HTTPException(status_code=400, detail=risk.reason)
    order = futu.place_order(req)
    return order


@router.get("/orders", response_model=list[Order])
async def get_orders():
    return futu.get_orders()

File: server/routes/portfolio.ts
import { Hono } from 'hono'
import { z } from 'zod'
import { zValidator } from '@hono/zod-validator'
import { getPositions, getBalance, getOrders, getOrder, checkRisk, placeOrder } from '../services/portfolio-db.js'

export const portfolioRouter = new Hono()

const orderBody = z.object({
  ticker: z.string().min(1),
  market: z.enum(['CN', 'US']),
  side: z.enum(['BUY', 'SELL']),
  qty: z.number().int().positive(),
  price: z.number().positive(),
  paper_trade: z.boolean().default(true),
})

portfolioRouter.get('/api/portfolio/positions', (c) => {
  return c.json(getPositions(Number(c.get('userId'))))
})

portfolioRouter.get('/api/portfolio/balance', (c) => {
  return c.json(getBalance(Number(c.get('userId'))))
})

portfolioRouter.get('/api/portfolio/orders', (c) => {
  return c.json(getOrders(Number(c.get('userId'))))
})

portfolioRouter.get('/api/portfolio/orders/:orderId', (c) => {
  const userId = Number(c.get('userId'))
  const orderId = c.req.param('orderId')
  const order = getOrder(userId, orderId)
  if (!order) return c.json({ detail: `Order '${orderId}' not found` }, 404)
  return c.json(order)
})

portfolioRouter.post('/api/portfolio/risk-check', zValidator('json', orderBody), (c) => {
  const req = c.req.valid('json')
  return c.json(checkRisk(Number(c.get('userId')), req))
})

portfolioRouter.post('/api/portfolio/orders', zValidator('json', orderBody), (c) => {
  const userId = Number(c.get('userId'))
  const req = c.req.valid('json')
  const risk = checkRisk(userId, req)
  if (!risk.passed) return c.json({ detail: risk.reason }, 422)
  const order = placeOrder(userId, req)
  return c.json(order, 201)
})
